src/gotcha_dl.c

```diff
@@ -5,7 +5,92 @@
 #include "elf_ops.h"
 #include <dlfcn.h>
 
-void* _dl_sym(void* handle, const char* name, void* where);
+/**
+ * structure used to pass lookup addr and return library address.
+ */
+struct Addrs {
+    ElfW(Addr) lookup_addr; // input
+    struct link_map *lmap;  // output
+};
+/**
+ * This is a callback to get headers for each library.
+ * We check if the caller's virtual address is between base address and the virtual addr + memory size.
+ * Then we select the lmap based on base_addr and name
+ *
+ * @param info, information about headers
+ * @param size, size of headers
+ * @param data, data from caller
+ * @return
+ */
+int lib_header_callback(struct dl_phdr_info * info, size_t size, void * data) {
+    struct Addrs* addrs = data;
+    char *name = NULL;
+    ElfW(Addr) load_address;
+    for (int i = 0; i < info->dlpi_phnum; ++i) {
+        if (info->dlpi_phdr[i].p_type == PT_LOAD) {
+            ElfW(Addr) base_addr = info->dlpi_addr + info->dlpi_phdr[i].p_offset;
```

Comment:
Still a wrong base_addr calculation. 'base_addr = info->dlpi_addr' is all you need.

p_offset tells you the byte offset of this PHDR in the on-disk library file. It's not part of an in-memory address calculation.

Reply:
fixed